### ISSUE TITLE
Storage: Directory structure

### ISSUE THREAD

**Issue by DanijelHuis** [aws-amplify/amplify-ios#358]
**Is your feature request related to a problem? Please describe.**
I am trying to upload a file using AWSS3StoragePlugin to a path like this:
_protected/city/poi/user_identity_id/photo.jpg_

but looking at the code, it seems like you are forcing a path to begin with: 
_protected/{user_identity_id}/_
as indicated here: https://aws-amplify.github.io/docs/js/storage#file-access-levels

For download I can override this behavior by providing targetIdentityId to StorageDownloadFileRequest. For upload on the other hand I can also provide
targetIdentityId to StorageUploadFileRequest but that property is not used anywhere.

**Describe the solution you'd like**
Allow us to pass whole serviceKey and not split it into access level, target identity id and key. 

**Describe alternatives you've considered**
```
AWSS3StorageUploadFileOperation (targetIdentityId should be passed here):
let serviceKey = StorageRequestUtils.getServiceKey(accessLevel: request.options.accessLevel, identityId: identityId, key: request.key)
```

**Additional context**
I can do all what I want by using transfer utility manually but I wanted to stick with AWSS3StoragePlugin.
Is there any special reason you are forcing _protected/{user_identity_id}/_? I am asking because S3 currently doesn't have a problem if I upload a file to different path.

**Comment by drochetti** [comment]
> Is there any special reason you are forcing protected/{user_identity_id}/? I am asking because S3 currently doesn't have a problem if I upload a file to different path.

@DanijelHuis no special reason other than a convention Amplify uses for user protected resources. That convention is true across all platforms (iOS, Android and JS) and allowing for custom paths would require a new feature to the added to the Storage API.

Are you blocked by that or is that something that would improve things on your side but not a blocker at this point?

**Comment by DanijelHuis** [comment]
We are not blocked, we moved the upload logic to API gateway so this is not needed for us anymore.
Thank you for your answer.

**Comment by drochetti** [comment]
No problem, thanks for getting back to me.

I'll close this for now since we don't intend to support it unless community interest is high across all supported platforms. If you have more details about how this could be useful let me know and I'm happy to open an RFC so the community can chime in.
